File: leetcode_go/200.Number_of_Islands.go
// https://leetcode.com/problems/number-of-islands/

var dir = [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

func numIslands(grid [][]byte) int {
    m := len(grid)
    n := len(grid[0])
    count := 0
    
    for i:=0;i<m;i++{
        for j:=0;j<n;j++{
            if grid[i][j] == '1' {
                count++
                dfs(&grid, i, j)
            }
        }
    }
    return count
}

func dfs(grid *[][]byte, row int, col int) {
    (*grid)[row][col] = '0'
    m := len(*grid)
    n := len((*grid)[0])
    
    for i:=0;i<4;i++ {
        r := row+dir[i][0]
        c := col+dir[i][1]
        if r >= 0 && r < m && c >= 0 && c < n && (*grid)[r][c] == '1' {
            dfs(grid, r, c)        
        }
    }
}

File: leetcode_cpp/200.Number_of_Islands.cpp
// https://leetcode.com/problems/number-of-islands/

class Solution {
 private:
  int m;
  int n;
  vector<pair<int, int>> dir =
      vector<pair<int, int>>{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

 public:
  int numIslands(vector<vector<char>>& grid) {
    m = grid.size();
    n = grid[0].size();
    int count = 0;

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (grid[i][j] == '1') {
          count++;
          dfs(grid, i, j);
        }
      }
    }

    return count;
  }

  void dfs(vector<vector<char>>& grid, int row, int col) {
    grid[row][col] = '0';

    for (int i = 0; i < 4; i++) {
      int r = row + dir[i].first;
      int c = col + dir[i].second;

      if (r >= 0 && r < m && c >= 0 && c < n && grid[r][c] == '1') {
        dfs(grid, r, c);
      }
    }
  }
};